qemu/tests/virtio_trace_pipenb.py

```diff
@@ -0,0 +1,42 @@
+import time
+from aexpect import ShellStatusError, ShellTimeoutError
+
+from avocado.utils import process
+from virttest import error_context
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Under named-pipe non-blocking testing:
+    1) Create pipe named by the following
+    2) Boot up a single-CPU guest with a virtio-serial device and
+       named-pipe chardev backend
+    3) Run a background work on the guest.
+    4) Write data to the virtio-serial port until the guest stops.
+    5) Read the named-pipe file on the host.
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment.
+    """
+
+    timeout = float(params.get("login_timeout", 360))
+    vm = env.get_vm(params["main_vm"])
+    serials = params["serials"].split()
+    v_path = vm.get_serial_console_filename(serials[-1])
+    vm.verify_alive()
+    session = vm.wait_for_login(timeout=timeout)
+    assert session.cmd_status_output(" while :;do sleep 1; date; done &") is not None, "There should be date in the output"
+    session.cmd_status_output("cat /proc/kallsyms > /dev/virtio-ports/vs2 &")
+    time.sleep(10)
+    out_put = process.getoutput("cat {}".format(v_path), timeout=timeout, shell=True)
```

Comment:
This command will be blocked until timeout since no EOF/EOL, 
Could you try:
```
    output = session.cmd_output("nohup cat /proc/kallsyms > /dev/virtio-ports/vs2 2>&1 &")
    guest_pid = output.split()[1]
    time.sleep(10)

    verify guest does not stuck 

    pipe = os.open(v_path, os.O_RDONLY | os.O_NONBLOCK)
    while True:
        try:
            os.read(pipe, 1)
        except OSError as e:
            if e.errno == errno.EAGAIN or e.errno == errno.EWOULDBLOCK:
                time.sleep(5)
                break
            else:
                raise Exception("Read data in host failed as %s" %e)

    if session.cmd_status("ps -p %s" %guest_pid):
        test.fail("send process in guest does not exit after all data are read out in host")
```

Reply:
Hi there is a problem with     if session.cmd_status("ps -p %s" %guest_pid):
due to the continous output for "while :;do sleep 1; date; done &"
It results in aexpect.exceptions.ExpectTimeoutError

Reply:
@michiboo  That does not make sense, `session.cmd_status("ps -p %s" %guest_pid)` is to get the status, the result is 1 is the process with guest_pid does not exist, you can get it from session log:
```
2020-06-08 21:57:08: [root@bootp-73-227-29 ~]#
2020-06-08 21:57:08: 1
```